File: apps/gateway/src/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";

export class CredentialCipher {
  constructor(
    private readonly key: Buffer,
    private readonly keyVersion = "v1",
  ) {
    if (key.length !== 32) throw new Error("Credential encryption key must be 32 bytes");
  }

  encrypt(value: unknown, associatedData: string): string {
    const iv = randomBytes(12);
    const cipher = createCipheriv("aes-256-gcm", this.key, iv);
    cipher.setAAD(Buffer.from(associatedData, "utf8"));
    const ciphertext = Buffer.concat([
      cipher.update(JSON.stringify(value), "utf8"),
      cipher.final(),
    ]);
    const tag = cipher.getAuthTag();
    return [this.keyVersion, iv.toString("base64url"), tag.toString("base64url"), ciphertext.toString("base64url")].join(".");
  }

  decrypt<T>(envelope: string, associatedData: string): T {
    const [version, ivPart, tagPart, ciphertextPart] = envelope.split(".");
    if (version !== this.keyVersion || !ivPart || !tagPart || !ciphertextPart) {
      throw new Error("Unsupported credential envelope");
    }
    const decipher = createDecipheriv("aes-256-gcm", this.key, Buffer.from(ivPart, "base64url"));
    decipher.setAAD(Buffer.from(associatedData, "utf8"));
    decipher.setAuthTag(Buffer.from(tagPart, "base64url"));
    const plaintext = Buffer.concat([
      decipher.update(Buffer.from(ciphertextPart, "base64url")),
      decipher.final(),
    ]);
    return JSON.parse(plaintext.toString("utf8")) as T;
  }
}

File: apps/gateway/polytrade_gateway/crypto.py
"""Authenticated encryption for custodial API credentials and alert targets."""

from __future__ import annotations

import base64
import json
import os
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _encode(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode().rstrip("=")


def _decode(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


class CredentialCipher:
    def __init__(self, key: bytes, key_version: str = "v1") -> None:
        if len(key) != 32:
            raise ValueError("Credential encryption key must be 32 bytes")
        self._cipher = AESGCM(key)
        self._key_version = key_version

    def encrypt(self, value: Any, associated_data: str) -> str:
        nonce = os.urandom(12)
        encrypted = self._cipher.encrypt(
            nonce,
            json.dumps(value, separators=(",", ":")).encode(),
            associated_data.encode(),
        )
        ciphertext, tag = encrypted[:-16], encrypted[-16:]
        return ".".join((self._key_version, _encode(nonce), _encode(tag), _encode(ciphertext)))

    def decrypt(self, envelope: str, associated_data: str) -> Any:
        parts = envelope.split(".")
        if len(parts) != 4 or parts[0] != self._key_version:
            raise ValueError("Unsupported credential envelope")
        _, nonce, tag, ciphertext = parts
        plaintext = self._cipher.decrypt(
            _decode(nonce),
            _decode(ciphertext) + _decode(tag),
            associated_data.encode(),
        )
        return json.loads(plaintext)
